src/preload/syscall_hook.S

```diff
@@ -572,6 +572,440 @@ SYSCALLHOOK_START(_syscall_hook_trampoline_49_89_ca)
+                        0x78, /* DW_OP_breg8 */         \
+                        (stub_scratch_2 - preload_thread_locals) | 0x80, \
+                        0x00
+        .cfi_escape     0x10, /* DW_CFA_expression */   \
+                        0x1e, /* x30 */                 \
+                        0x03, /* length 3 */            \
+                        0x78, /* DW_OP_breg8 */         \
+                        (stub_scratch_2 - preload_thread_locals + 8) | 0x80, \
+                        0x00
+        .cfi_escape     0x16, /* DW_CFA_val_expression */                       \
+                        0x08, /* x8 */                                          \
+                        0x06, /* length 6 */                                    \
+                        0x78, /* DW_OP_breg8 */                                 \
+                        ((-preload_thread_locals) & 0x7F) | 0x80,               \
+                        ((-preload_thread_locals) & (0x7F << 7)) >> 7 | 0x80,   \
+                        ((-preload_thread_locals) & (0x7F << 14)) >> 14 | 0x80, \
+                        ((-preload_thread_locals) & (0x7F << 21)) >> 21 | 0x80, \
+                        ((-preload_thread_locals) & (0x7F << 28)) >> 28 | 0x70
+        mov     x15, preload_thread_locals
+        // Stash away x30 so that we can have two registers to use again
+        // we can't use stub_scratch_2 since we might overwrite the data there
+        str     x30, [x15, stub_scratch_1 - preload_thread_locals]
+        .cfi_escape     0x10, /* DW_CFA_expression */                   \
+                        0x20, /* pc */                                  \
+                        0x06, /* length 6 */                            \
+                        0x0c, /* DW_OP_const4u */                       \
+                        stub_scratch_1 & 0xFF,                          \
+                        (stub_scratch_1 & (0xFF <<  0x8)) >>  0x8,      \
+                        (stub_scratch_1 & (0xFF << 0x10)) >> 0x10,      \
+                        (stub_scratch_1 & (0xFF << 0x18)) >> 0x18,      \
+                        0x06, /* DW_OP_deref */
+        // Move the saving area to the start of scratch_2
+        // Do it in the forward order since we know x8 >= x15
+        ldr     x30, [x8, stub_scratch_2 - preload_thread_locals]
+        str     x30, [x15, stub_scratch_2 - preload_thread_locals]
+        ldr     x30, [x8, stub_scratch_2 - preload_thread_locals + 8]
+        str     x30, [x15, stub_scratch_2 - preload_thread_locals + 8]
+        // Restore x8
+        movk    x8, 0, lsl 16
+        .cfi_restore x8
+        .cfi_escape     0x10, /* DW_CFA_expression */                   \
+                        0x0f, /* x15 */                                 \
+                        0x05, /* length 5 */                            \
+                        0x0c, /* DW_OP_const4u */                       \
```

Comment:
I think at this point this pattern is used often enough to warrant a macro to encode this.